rewrite-java-test/src/test/java/org/openrewrite/java/cleanup/RemoveInstanceOfPatternMatchTest.java

```diff
@@ -0,0 +1,880 @@
+              """,
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String && (String) obj != null) {
+                          String str = (String) obj;
+                          System.out.println(str);
+                      }
+                  }
+              }
+              """),
+            14));
+    }
+
+    @Test
+    void unusedVariable() {
+        rewriteRun(
+          version(
+            java(
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String str) {
+                          System.out.println();
+                      }
+                  }
+              }
+              """,
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String) {
+                          System.out.println();
+                      }
+                  }
+              }
+              """),
+            14));
+    }
+
+    @Test
+    void disjunction() {
+        rewriteRun(
+          version(
+            java(
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String str && str.isEmpty() || false) {
+                          System.out.println();
+                      }
+                  }
+              }
+              """,
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String && ((String) obj).isEmpty() || false) {
+                          System.out.println();
+                      }
+                  }
+              }
+              """),
+            14));
+    }
+
+    @Test
+    void localVariable() {
+        rewriteRun(
+          version(
+            java(
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String str || true) {
+                          String str = null;
+                          System.out.println(str);
+                      }
+                  }
+              }
+              """,
+              """
+              package com.example;
+
+              class Example {
+                  public void test(Object obj) {
+                      if (obj instanceof String || true) {
+                          String str = null;
+                          System.out.println(str);
+                      }
+                  }
+              }
+              """),
+            14));
+    }
+
+    @Test
+    @Disabled("Not supported")
```

Comment:
There is one more case here which at some point may also have to be covered:
```java
    public void test(Object obj) {
        if (!(obj instanceof String str)) {
            throw new IllegalArgumentException();
        }
        System.out.println(str.length());
    }
```
I've never seen code like this in the wild, but it is actually legal.

Reply:
I added a test case and disabled it. I didn't know that it's correct to use a variable outside if/else. I think It can be implemented, but it requires some rework.

Actually the possibility to navigate from variable usage to its declaration would simplify the implementation. J.Identifier::getFieldType() contains a lot of useful information on a variable. The only missing thing is a reference to a declaring expression.